ports/espressif/common-hal/wifi/__init__.c

```diff
@@ -104,14 +104,19 @@ static void event_handler(void *arg, esp_event_base_t event_base,
                 wifi_event_sta_disconnected_t *d = (wifi_event_sta_disconnected_t *)event_data;
                 uint8_t reason = d->reason;
                 ESP_LOGW(TAG, "reason %d 0x%02x", reason, reason);
-                if (radio->retries_left > 0 &&
-                    reason != WIFI_REASON_AUTH_FAIL &&
-                    reason != WIFI_REASON_NO_AP_FOUND &&
-                    reason != WIFI_REASON_ASSOC_LEAVE) {
-                    radio->retries_left--;
-                    ESP_LOGI(TAG, "Retrying connect. %d retries remaining", radio->retries_left);
-                    esp_wifi_connect();
-                    return;
+                if (radio->retries_left > 0) {
+                    if (
+                        reason != WIFI_REASON_AUTH_FAIL &&
+                        reason != WIFI_REASON_NO_AP_FOUND &&
+                        reason != WIFI_REASON_ASSOC_LEAVE) {
+                        radio->retries_left--;
+                        ESP_LOGI(TAG, "Retrying connect. %d retries remaining", radio->retries_left);
+                        esp_wifi_connect();
+                        return;
+                    }
+                } else if (reason == WIFI_REASON_4WAY_HANDSHAKE_TIMEOUT) {
+                    ESP_LOGW(TAG, "reason %d masked as %d", WIFI_REASON_4WAY_HANDSHAKE_TIMEOUT, WIFI_REASON_AUTH_FAIL);
+                    reason = WIFI_REASON_AUTH_FAIL;
                 }
```

Comment:
I don't think it makes sense to map the reason to something else here because you lose information. Instead, do this mapping where the reason is used to produce an error.

Reply:
The thing is that the idf now never returns `WIFI_REASON_AUTH_FAIL`, only `WIFI_REASON_4WAY_HANDSHAKE_TIMEOUT`.
So information is not lost by the code, but by the idf that broke `WIFI_REASON_AUTH_FAIL`.
We are now always getting `WIFI_REASON_4WAY_HANDSHAKE_TIMEOUT`.

So at least in the current version of the idf, we cannot tell what `WIFI_REASON_4WAY_HANDSHAKE_TIMEOUT` really means, and we have to use a single error code.

Reply:
@anecdata I'm not sure: could you clarify whether you think the proposed PR is what you'd like, or you want it to do something else? Thanks.

Reply:
There are two distinct situations.

1. Connect to secure wifi w/o password:

Here, behavior doesn't match Espressif docs: https://docs.espressif.com/projects/esp-idf/en/v5.1.3/esp32s2/api-guides/wifi.html#wi-fi-reason-code-related-to-wrong-password I'd expect a `201` (`NO_AP_FOUND`). The docs vary from the behavior we see in other ways too, so 🤷‍♂️ . After more review, I don't think the API has the granularity where we could detect that case in code and pre-empt a connection, short of adding our own scan which is not ideal.

2. Connect to secure wifi with incorrect password:

Current docs (link above) do say that
> Four-way handshake times out. Setting wrong password when STA connecting to an encrypted AP

should give a `15` (`4WAY_HANDSHAKE_TIMEOUT`), which it does.

 I don't have strong preferences here. With either error, the connection failed and users are going to fidget with possibilities. We don't know Espressif's intent with the change(s), or why the docs don't seem to match reality in all cases. There's never really been a clear description of how (or why) the Espressif-specific 2xx codes differ from the corresponding 1- and 2-digit IEEE codes. In some cases (per Espressif docs) the error code is coming from the AP. IEEE code `15` / `4WAY_HANDSHAKE_TIMEOUT` yields web search results about password issues. For the 2xx codes, we're at the mercy of Espressif's docs.

Perhaps the best solution is to consider in the future adding descriptive strings for the more common error codes (I seem to recall some discussion of this).

Reply:
To be clearer. I'm ok with this remapping but want it to happen outside this function.

Specifically, change how it works here: https://github.com/adafruit/circuitpython/blob/main/ports/espressif/common-hal/wifi/Radio.c#L379-L383 You can change what triggers the AUTH_FAIL to python there.

Reply:
> short of adding our own scan which is not ideal

@anecdata I mean, wifi is already taking 95% of boot time of pretty much everything.
It already feels like a system hang. One more thing isn't gonna make a difference.

> Perhaps the best solution is to consider in the future adding descriptive strings for the more common error codes (I seem to recall some discussion of this).

Error codes should be accompanied with links. This is the one thing Windows of all things does really well with powershell.
If there were to be a function that prints an exception, takes the changing part of a url, prints the static url and then seperately the changing part, this could be done space efficiently.

> Specifically, change how it works here

@tannewt Did so, and tested on s2.